### DISCUSSION TITLE
Ability to resolve route models from the form request

### DISCUSSION THREAD

**Discussion by MatanYadaev** [laravel/framework#38432] [Ideas]
It happened many times that in my company we wanted to use route model bindings but didn't have the option to control it easily. [Customizing the resolution logic](https://laravel.com/docs/8.x/routing#customizing-the-resolution-logic) in the service provider is fine for cases you re-use the model, but it doesn't happen a lot, and it gives more control, less coupling, and better readability to have the resolution within the form request.

So in my company we prefer doing this today (as a workaround):
```php
<?php

Route::put('posts/{id}', [PostController::class, 'update']);

class PostController
{
  public function update($id, UpdatePostRequest $request)
  {
    $post = $request->post();
    
    //
  }
}

class UpdatePostRequest extends FormRequest
{
  public function post(): Post
  {
    // Resolve only once
    return once(function(): Post {
      return Post::withTrashed()
        ->select([...])
        ->whereHas(...)
        ->with([...])
        ->findOrFail($this->route('id'));
    });
  }

  public function authorize(): bool
  {
    return $this->user()->id === $this->post()->user_id;
  }
}
```

It could be better to have this supported:
```php
<?php

Route::put('posts/{post}', [PostController::class, 'update']);

class PostController
{
  public function update(Post $post, UpdatePostRequest $request)
  {
    //
  }
}

class UpdatePostRequest extends FormRequest
{
  public function resolvePost($id): Post
  {
    return Post::withTrashed()
      ->select([...])
      ->whereHas(...)
      ->with([...])
      ->findOrFail($id);
  }

  public function authorize(): bool
  {
    // `$this->route('post') will look for `$this->resolvePost` and
    // as a fallback will use the explicit/implicit model bindings that configured in the service provider
    return $this->user()->id === $this->route('post')->user_id;
  }
}
```

**Comment by spaceworkplatform** [comment]
Agree, there are times that we also want to control the model binding for specific request.
Could be very helpful.

**Comment by hananbo** [comment]
it's nice and clean! it can also reduce redundant duplicated queries done afterwards in the controller.

**Comment by zserhiy** [comment]
I use similar feature but method resolvePost moved to model.
At the moment, I enhanced SubstituteBindings by writing descendant.
I combine implicit bindings via dependency injection with "explicit" binding for loading model.
Your route will look like
$router->put('posts/{post}', [PostController::class, 'update'])->middleware('**bindings:post@resolvePost**');
Routines for loading are compatible with UrlRoutable by parameters
So I can write functions like loadWithTrashed, loadOnlyTrashed and so on.
I've already seen new Laravel feature withTrashed in routing. But this feature is not multipurpose unlike my variant and even to yours.

**Comment by erikgaal** [comment]
If the model is resolved with route model binding, simply using `request()->route('post')` will return the Model instance.
